src/kernels/backend_webgl.ts

```diff
@@ -1630,6 +1630,55 @@ export class MathBackendWebGL implements KernelBackend {
     return this.compileAndRun(program, [x]) as T;
   }
 
+  conv2dByMatMul(x: Tensor4D, filter: Tensor4D, convInfo: Conv2DInfo):
+      Tensor4D {
+    // Reshapes conv2D input to 2D tensors, uses matMul and then reshape the
+    // result from 2D to 4D.
+    if (!ENV.get('WEBGL_LAZILY_UNPACK') ||
+        !ENV.get('WEBGL_PACK_BINARY_OPERATIONS')) {
+      const $x = this.reshape(x, [1, x.shape[0] * x.shape[1] * x.shape[2],
+          convInfo.inChannels]) as Tensor3D;
+      const $filter = this.reshape(filter,
+          [1, convInfo.inChannels, convInfo.outChannels]) as Tensor3D;
+      return this.batchMatMul($x, $filter, false, false).reshape<Rank.R4>(
+          convInfo.outShape);
+    }
+    
+    const xshape = x.shape;
+    const xTexData = this.texData.get(x.dataId);
+    // We avoid expensive packed reshape by padding row count to next even
+    // number. When number of rows in batch, x.shape[2]), in 2x2 packed texture
+    // layout is odd, the result of packed batchMatMul is the same as it is for
+    // next, even, number of rows in the batch - x.shape[2] + 1.
+    const packedXRowPad = xshape[2] % 2 === 1 && xTexData.isPacked ? 1 : 0;
+
+    const $x = Tensor.make(
+        [1, xshape[0] * xshape[1] * (xshape[2] + packedXRowPad),
+         convInfo.inChannels], {dataId: x.dataId}, x.dtype) as Tensor3D;
+    if (packedXRowPad) {
+        xTexData.shape[xTexData.shape.length - 2]++;
+        util.assert(webgl_util.isReshapeFree(xTexData.shape, $x.shape),
+            `packed reshape from ${xTexData.shape} to ${$x.shape} isn't free`,);
+    }
+    const $filter = this.reshape(filter,
+        [1, convInfo.inChannels, convInfo.outChannels]) as Tensor3D;
+    
+    const xf = this.batchMatMul($x, $filter, false, false);
```

Comment:
What does `xf` stand for? Could you rename this to something a little more descriptive, like `pointwiseConv`?